### ISSUE TITLE
ng-content, slot and content elements don't work with nested Angular custom elements

### ISSUE THREAD

**Issue by JenPeleva** [angular/angular#24536]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
Currently, when you have nested Angular custom elements, and the child element contains ng-content,  content or slot elements in its template, content projection/insertion points don't work. Here's a more detailed explanation of the behavior:

 For the sake of the example, let's have two Angular components - container and button. Button uses content projection and the template of the container component has nested button inside it. 

Tried all kinds of scenarios with nested custom elements - using the content element from v.0 of ShadowDom, the slot from v. 1 and ng-content and placing them in the app.component.html, in the angular html page in the src folder, as well as on a separately hosted html page, where I'm loading my elements bundles, generated after a production build (following this approach: https://medium.com/@tomsu/building-web-components-with-angular-elements-746cd2a38d5b).

Here's what I noticed:

## 'slot' and 'content' elements behave similarly
1. Insertion points work for both the button and the container, when they are placed in the app.component's template.
2. In the angular src/html - the button's content doesn't show up. The content of the button in the container shows up.
3. On the separately hosted html page - same behavior as 2.

You can check a demo with slots here: https://stackblitz.com/edit/angular-6jmqbg

## ng-content
1. Content projection doesn't work when the button and the container are placed in the app.component.
2. On the src/html page - ng-content works for the button, but doesn't work in the container.
3. In the external html - same as 2 

You can check a demo project with ng-content here: https://stackblitz.com/edit/angular-uxthfz

## Expected behavior
Shadow dom insertion points or/and content projection should work in nested Angular custom elements

## Minimal reproduction of the problem with instructions
You can check a demo with slots here: https://stackblitz.com/edit/angular-6jmqbg
You can check a demo project with ng-content here: https://stackblitz.com/edit/angular-uxthfz

## What is the motivation / use case for changing the behavior?

I'm mostly interested in the last case - in which I'm loading the bundled elements on a a separately hosted html page (external application). It's pretty blocking for us, because most of our components are complex and contain other components in their templates. 

Thanks, guys!  :cake:

## Environment

<pre><code>
Angular version: 6.0.0

**Comment by benoitjchevalier** [comment]
I was also fiddling with the same scenario for the past couple of days and about to open an issue. I've only played with `ng-content` as I was assuming that angular-element would taker care of using `content` behind the scene in a custom element scenario. Because angular isn't using Shadow Dom v1 I'd be surprised if we could use `slot`.

What I noticed is that when nesting those `ng-content` the content is sometimes projected as a sibling of the shadow-root in the child component, hence the fact that the nested angular element `ng-content` then doesn't re-project it. in example:

Given
outer component: `<inner><ng-content></ng-content></inner>`
inner component: `<div><ng-content></ng-content></div>`

this code
```
<outer>
   <div>Project me!</div>
</outer>
```

results in this DOM distribution:
<outer>
   #shadow-root
       <inner>
              #shadow-root
                     <div></div>
              <div>Project me!</div>
        </inner>
</outer>

hence the `inner` `ng-content` can't select the `Project me!` div and redistribute it properly.


For what it's worth I've noticed that custom-element registration order seem to have an impact. If you register `outer` first then the DOM distribution is fine. It seems counter-intuitive as `outer` uses `inner`, I would have assumed `inner` needed to be registered first. @JenPeleva this might be why you're seeing a difference between the two examples in the app and directly in the HTML? Not sure in what order the registration happens in your scenario.
However if you register `outer` first then it seems like other custom elements within `outer` don't get their template dynamically updated with changes to their `@Input()` anymore.

Example of registering `inner` first, DOM distribution doesn't work but `@Input()` props in template get updated: https://stackblitz.com/edit/angular-custom-element-order
Example of registering `outer` first, templates are updated but DOM distribution is wrong: https://stackblitz.com/edit/angular-custom-element-order2

This is also a very important feature for us as we are building a library of web components using Angular Elements. Maybe this could be looked at when doing the new slot feature in 
#23636 ?

**Comment by mhevery** [comment]
@robwormald Can you look into this more in depth, but I think the short of it is that it is expected. It may be fixed by ivy?

**Comment by robwormald** [comment]
Not really fixed by Ivy, but it’s tracked here : https://github.com/angular/angular/issues/23636 - there’s a PR in flight. 

Tldr: `<ng-content>` works in a very limited fashion with Elements (basically, only for very static stuff) and won’t provide any support for `@ContentChild/ren` queries (because they have to be known at compile time)

The pending PR adds support for Shadow DOM v1, which uses `<slot>` elements to allow content projection. See https://webkit.org/blog/4096/introducing-shadow-dom-api/ for a good explainer on this API. 

Again, this won’t yet support queries, but developers can use the `<slot (slotchange)>` event to approximate this behavior and allow wiring dynamic stuff up.

**Comment by mhevery** [comment]
@robwormald So it sounds like this should be closed with won't fix.

**Comment by robwormald** [comment]
Closing this as:
- wontfix for <ng-content> 
- resolved for <slot> via #24861
